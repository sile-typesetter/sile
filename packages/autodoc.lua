--
-- Documentation tooling for package designers.
--

SILE.registerCommand("package-documentation", function (options, _)
  local package = SU.required(options, "src", "src for package documentation")
  SU.debug("autodoc", package)
  local exports = require(package)
  if type(exports) ~= "table" or not exports.documentation then
    SU.error("Undocumented package "..package)
  end
  SILE.process(
    SILE.inputs.TeXlike.docToTree(
      exports.documentation
    )
  )
end)

-- Styling hook

SILE.settings.declare({
  parameter = "autodoc.highlighting",
  default = false,
  type = "boolean",
  help = "Whether audodoc enables syntax highlighting"
})

SILE.scratch.autodoc = {
  theme = {
    command = "#1d4851", -- oil blue
    parameter = "#3f5218", -- some sort of dark green
    setting = "#42280e", -- some kind of dark brown
    bracketed = "#656565", -- some grey
    package = "#172557", -- saturated space blue
  }
}
local colorWrapper = function (ctype, content)
  local color = SILE.scratch.autodoc.theme[ctype]
  if color and SILE.settings.get("autodoc.highlighting") and SILE.Commands["color"] then
    SILE.call("color", { color = color }, content)
  else
    SILE.process(content)
  end
end

SILE.registerCommand("autodoc:package:style", function (_, content)
  SILE.call("font", { weight = 700 }, function()
    colorWrapper("package", content)
  end)
end)

SILE.registerCommand("autodoc:code:style", function (options, content)
  -- options.type is used to distinguish the type of code element and style
  -- it accordingly: "ast", "setting", "environment" shall select the font
  -- (by default, using \code) and color, the other (lower-level in an AST)
  -- shall select only the color.
  if options.type == "ast" then
    SILE.call("code", {}, content)
  elseif options.type == "setting" then
    SILE.call("code", {}, function()
      colorWrapper(options.type, content)
    end)
  elseif options.type == "environment" then
    SILE.call("code", {}, function()
      colorWrapper("command", content)
    end)
  else
    colorWrapper(options.type, content)
  end
end)

-- Documenting a setting with good line-breaks

local inputfilter = SILE.require("packages/inputfilter").exports
local settingFilter = function (node, content)
  if type(node) == "table" then return node end
  local result = {}
  for token in SU.gtoke(node, "[%.]") do
    if token.string then
      result[#result+1] = token.string
    else
        result[#result+1] = token.separator
        result[#result+1] = inputfilter.createCommand(
          content.pos, content.col, content.line,
          "penalty", { penalty = 100 }, nil
        )
    end
  end
  return result
end

SILE.registerCommand("autodoc:setting", function (options, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  if #content ~= 1 then SU.error("Expected a single element") end
  local name = type(content[1] == "string") and content[1]
  if not name then SU.error("Unexpected setting") end
  -- Conditional existence check (can be disable is passing check=false), e.g.
  -- for settings that would be define in another context.
  if SU.boolean(options.check, true) then
    SILE.settings.get(name) -- will issue an error if unknown
  end
  -- Inserts breakpoints after dots
  local nameWithBreaks = inputfilter.transformContent(content, settingFilter)

  SILE.call("autodoc:code:style", { type = "setting" }, nameWithBreaks)
end, "Outputs a settings name in code, ensuring good line breaks and possibly checking their existence.")

-- Documenting a command, benefiting from AST parsing

local function optionSorter(o1, o2)
  -- options are in an associative table and Lua doesn't guarantee a fixed order.
  -- To ensure we get a consistent and stable output, we make with some wild guesses here
  -- (Quick'n dirty, could be improved!), and rely on alphabetical order otherwise.
  if o1 == "src" then return true end
  if o2 == "src" then return false end
  if o1 == "name" then return true end
  if o2 == "name" then return false end
  return o1 < o2
end

local function typesetAST(options, content)
  if not content then return end
  local seenCommandWithoutArg = false
  for i = 1, #content do
    local ast = content[i]

    if type(ast) == "string" then
      if seenCommandWithoutArg and ast:sub(1,1) ~= " " and ast:sub(1,1) ~= "{" then
        -- Touchy:
        -- There might have been a space or a {} here in the original code. The AST does
        -- not remember it, we only know we have to separate somehow the string from
        -- the previous command...
        SILE.typesetter:typeset(" ")
        seenCommandWithoutArg = false
      end
      if ast:sub(1, 1) == "<" and ast:sub(-1) == ">" then
        SILE.call("autodoc:internal:bracketed", {}, { ast:sub(2, -2) })
      else
        SILE.typesetter:typeset(ast)
      end

    elseif ast.command then
      local cmd = SILE.Commands[ast.command]
      if not cmd and SU.boolean(options.check, true) then
        SU.error("Unexpected command '"..ast.command.."'")
      end
      SILE.typesetter:typeset("\\")
      SILE.call("autodoc:code:style", { type = "command" }, { ast.command })

      local sortedOpts = {}
      for k, _ in pairs(ast.options) do table.insert(sortedOpts, k) end
      table.sort(sortedOpts, optionSorter)
      if #sortedOpts > 0 then
        SILE.typesetter:typeset("[")
        for iOpt, option in ipairs(sortedOpts) do
          SILE.call("autodoc:code:style", { type = "parameter" }, { option })
          SILE.typesetter:typeset("=")
          SILE.call("penalty", { penalty = 100 }, nil) -- Quite decent to break here if need be.
          SILE.call("autodoc:value", {}, { ast.options[option] })
          if iOpt == #sortedOpts then
            SILE.typesetter:typeset("]")
          else
            SILE.typesetter:typeset(", ")
          end
        end
      end
      if (#ast >= 1) then
        SILE.call("penalty", { penalty = 200 }, nil) -- Less than optimal break.
        SILE.typesetter:typeset("{")
        typesetAST(options, ast)
        SILE.typesetter:typeset("}")
      else
        seenCommandWithoutArg = true
      end

    else
      SU.error("Unrecognized AST element")
    end
  end
end

SILE.registerCommand("autodoc:internal:ast", function (options, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  SILE.call("autodoc:code:style", { type = "ast" }, function ()
    typesetAST(options, content)
  end)
end, "Outputs a nicely typeset AST (low-level command).")

SILE.registerCommand("autodoc:internal:bracketed", function (_, content)
  SILE.typesetter:typeset("⟨")
  SILE.call("autodoc:code:style", { type = "bracketed" }, function()
    SILE.call("em", {}, content)
  end)
  SILE.call("kern", { width = "0.1em" }) -- fake italic correction.
  SILE.typesetter:typeset("⟩")
end, "Outputs a nicely formatted user-given value within <brackets>.")

SILE.registerCommand("autodoc:value", function (_, content)
  local value = type(content) == "table" and content[1] or content
  if type(value) ~= "string" then SU.error("Expected a string") end

  if value:sub(1, 1) == "<" and value:sub(-1) == ">" then
    SILE.call("autodoc:internal:bracketed", {}, { value:sub(2, -2) })
  else
    -- Here we should check for comma/semicolon, or surrounding spaces, and in that
    -- case add quotes around the value. This is a bit of an edge-case though, esp.
    -- for documentation needs.
    SILE.call("autodoc:code:style", { type = "value" }, content)
  end
end, "Outputs a nicely formatted argument within <brackets>.")

SILE.registerCommand("autodoc:command", function (options, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  if type(content[1]) ~= "table" then SU.error("Expected a command, got "..type(content[1]).." '"..content[1].."'") end

  SILE.call("autodoc:internal:ast", options, content)
end, "Outputs a formatted command, possibly checking its validity.")

-- Documenting a parameter

SILE.registerCommand("autodoc:parameter", function (_, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  if #content ~= 1 then SU.error("Expected a single element") end
  local param = type(content[1] == "string") and content[1]

  local parts = {}
  for v in string.gmatch(param, "[^=]+") do
    parts[#parts+1] = v
  end
  SILE.call("autodoc:code:style", { type = "ast" }, function ()
    if #parts < 1 or #parts > 2 then SU.error("Unexpected parameter '"..param.."'") end
    SILE.call("autodoc:code:style", { type = "parameter" }, { parts[1] })
    if #parts == 2 then
      SILE.typesetter:typeset("=")

      SILE.call("penalty", { penalty = 100 }, nil) -- Quite decent to break here if need be.
      SILE.call("autodoc:value", {}, { parts[2] })
    end
  end)
end, "Outputs a nicely presented parameter, possibly with a value.")

-- Documenting an environment

SILE.registerCommand("autodoc:environment", function (options, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  if #content ~= 1 then SU.error("Expected a single element") end
  local name = type(content[1] == "string") and content[1]
  if not name then SU.error("Unexpected environment") end
  -- Conditional existence check
  if SU.boolean(options.check, true) then
    if not SILE.Commands[name] then SU.error("Unknown command "..name) end
  end

  SILE.call("autodoc:code:style", { type = "environment" }, name)
end, "Outputs a command name in code, checking its validity.")

-- Documenting a package name

SILE.registerCommand("autodoc:package", function (_, content)
  if type(content) ~= "table" then SU.error("Expected a table content") end
  if #content ~= 1 then SU.error("Expected a single element") end
  local name = type(content[1] == "string") and content[1]
  if not name then SU.error("Unexpected package name") end
  -- We cannot really check package name to exist!

  SILE.call("autodoc:package:style", {}, { name })
end, "Outputs a package name in code, checking its validity.")

return {
  documentation = [[\begin{document}
This package extracts documentation information from other packages. It’s used to
construct the SILE manual. Keeping package documentation in the package itself
keeps the documentation near the implementation, which (in theory) makes it easy
for documentation and implementation to be in sync.

For that purpose, it provides the \autodoc:command{\package-documentation[src=<package>]}
command.

Properly documented packages should export a \code{documentation} string
containing their documentation, as a SILE document.

For documenters and package authors, it also provides commands that can be used in their package
documentation to present various pieces of information in a consistent way.

Setting names can be fairly long (e.g. \em{namespace.area.some-stuff\kern[width=0.1em]}).
The \autodoc:command{\autodoc:setting} command helps line-breaking them automatically at
appropriate points, so that package authors do not have care about them
manually.

With the \autodoc:command{\autodoc:command} command, one can pass a simple command, or even
a full commands (with parameters and arguments), without the need for escaping special
characters. This relies on SILE’s AST (abstract syntax tree) parsing, so you benefit from
typing simplicity, syntax check, and even more –such as styling\footnote{If the \autodoc:package{color}
package is loaded and the \autodoc:setting{autodoc.highlighting} setting is set to true, you get syntax
highlighting.}.
Moreover, for text content in parameter values or command arguments, if they are enclosed
between angle brackets, they will be presented with an distinguishable style.
Just type the command as it would appear in code, and it will be nicely typeset. It comes with
a few caveats, though: parameters are not guaranteed to appear in the order you entered them and
some purely syntactic sequences are just skipped and not reconstructed. Also, it is not adapted
to math-related commands. So it comes with many benefits, but also at a cost.

The \autodoc:command{\autodoc:environment} command just takes an environment name, so
basically a command, but just displays it without leading backslash.

The \autodoc:command{\autodoc:setting}, \autodoc:command{\autodoc:command} and
\autodoc:command{\autodoc:environment} commands all check the validity and existence of
their inputs. Would you want to disable this feature (e.g. to refer to a setting or command
defined in another package or module that might not be loaded at this point), you can set the
optional parameter \autodoc:parameter{check} to false.
Note, however, that for commands, it is applied recursively to the parsed AST
(so it is a all-or-none trade-off).

The \autodoc:command{\autodoc:parameter} commands takes either a parameter name, possibly
with a value (which as above, may be bracketed) and typesets it in the same fashion.

\end{document}]]
}
