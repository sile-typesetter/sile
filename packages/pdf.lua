if SILE.outputter ~= SILE.outputters.libtexpdf then
  SU.error("pdf package requires libtexpdf backend")
end
local pdf = require("justenoughlibtexpdf")

SILE.registerCommand("pdf:destination", function (options, content)
  local name = SU.required(options, "name", "pdf:bookmark")
  SILE.typesetter:pushHbox({
    value = nil,
    height = 0,
    width = 0,
    depth = 0,
    outputYourself = function (self, typesetter)
      SILE.outputters.libtexpdf._init()
      pdf.destination(name, typesetter.frame.state.cursorX, SILE.documentState.paperSize[2] - typesetter.frame.state.cursorY)
    end
  })
end)

SILE.registerCommand("pdf:bookmark", function (options, content)
  local dest = SU.required(options, "dest", "pdf:bookmark")
  local title = SU.required(options, "title", "pdf:bookmark")
  local level = options.level or 1
  -- Added UTF8 to UTF16-BE conversion
  -- For annotations and bookmarks, text strings must be encoded using
  -- either PDFDocEncoding or UTF16-BE with a leading byte-order marker.
  -- As PDFDocEncoding supports only limited character repertoire for
  -- European languages, we use UTF-16BE for internationalization.
  local ustr = SU.utf8_to_utf16be_hexencoded(title)
  SILE.typesetter:pushHbox({
    value = nil, height = 0, width = 0, depth = 0,
    outputYourself = function ()
      local d = "<</Title<" .. ustr .. ">/A<</S/GoTo/D(" .. dest .. ")>>>>"
      SILE.outputters.libtexpdf._init()
      pdf.bookmark(d, level)
    end
  })
end)

if SILE.Commands.tocentry then
  SILE.scratch.pdf = { dests = {}, dc = 1 }
  local oldtoc = SILE.Commands.tocentry
  SILE.Commands.tocentry = function (options, content)
    SILE.call("pdf:destination", { name = "dest" .. SILE.scratch.pdf.dc } )
    local title = SU.contentToString(content)
    SILE.call("pdf:bookmark", { title = title, dest = "dest" .. SILE.scratch.pdf.dc, level = options.level })
    oldtoc(options, content)
    SILE.scratch.pdf.dc = SILE.scratch.pdf.dc + 1
  end
end

SILE.registerCommand("pdf:literal", function (options, content)
  SILE.typesetter:pushHbox({
      value = nil,
      height = 0,
      width = 0,
      depth = 0,
      outputYourself = function (self, typesetter)
        SILE.outputters.libtexpdf._init()
        pdf.add_content(content[1])
      end
    })
end)

SILE.registerCommand("pdf:link", function (options, content)
  local dest = SU.required(options, "dest", "pdf:link")
  local target = options.external and "/Type/Action/S/URI/URI" or "/S/GoTo/D"
  local llx, lly
  SILE.typesetter:pushHbox({
    value = nil,
    height = 0,
    width = 0,
    depth = 0,
    outputYourself = function (self,typesetter)
      llx = typesetter.frame.state.cursorX
      lly = SILE.documentState.paperSize[2] - typesetter.frame.state.cursorY
      SILE.outputters.libtexpdf._init()
      pdf.begin_annotation()
    end
  })

  local hbox = SILE.Commands["hbox"]({}, content) -- hack

  SILE.typesetter:pushHbox({
    value = nil,
    height = 0,
    width = 0,
    depth = 0,
    outputYourself = function (self,typesetter)
      local d = "<</Type/Annot/Subtype/Link/C [ 1 0 0 ]/A<<" .. target .. "(" .. dest .. ")>>>>"
      pdf.end_annotation(d, llx, lly, typesetter.frame.state.cursorX, SILE.documentState.paperSize[2] -typesetter.frame.state.cursorY + hbox.height)
    end
  })
end)

return { documentation = [[\begin{document}
The \code{pdf} package enables (basic) support for PDF links and table-of-contents
entries. It provides the three commands \command{\\pdf:destination}, \command{\\pdf:link}
and \command{\\pdf:bookmark}.

The \command{\\pdf:destination} parameter creates a link target; it expects a
parameter called \code{name} to uniquely identify the target. To create a link to
that location in the document, use \code{\\pdf:link[dest=\goodbreak{}name]\{link content\}}.

If the \code{pdf} package is loaded after the \code{tableofcontents} package (e.g.
in a document with the \code{book} class), then a PDF document outline will be generated.
\end{document}]] }
