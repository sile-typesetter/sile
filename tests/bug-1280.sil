\begin[papersize=15cm x 6cm]{document}
\font[filename=.fonts/NotoSerifCJK-Regular.ttc]
\nofolios
\neverindent
\script{
  local checks = 0;
  local function check() SILE.typesetter:typeset("‚úì"); checks = checks + 1 end
  function test(teststr)
  SILE.call("par")
  SILE.typesetter:typeset(teststr .. " ‚áí")
  local function decbe(s) return SU.utf16be_to_utf8(SU.hexdecoded(s)) end
  local function decle(s) return SU.utf16le_to_utf8(SU.hexdecoded(s)) end

  SILE.call("par")
  local be = SU.utf8_to_utf16be_hexencoded(teststr)
  SILE.typesetter:typeset(be .. " ‚Üí ")
  if string.sub(be,1,4) == "feff" then check() end
  SILE.typesetter:typeset(decbe(be))
  SILE.call("par")
  be = string.sub(be, 5, #be)
  SILE.typesetter:typeset(be .. " ‚Üí ")
  if decbe(be) == teststr then check() end
  SILE.typesetter:typeset(decbe(be))

  SILE.call("par")
  local le = SU.utf8_to_utf16le_hexencoded(teststr)
  SILE.typesetter:typeset(le .. " ‚Üí ")
  if string.sub(le,1,4) == "fffe" then check() end
  SILE.typesetter:typeset(decle(le))
  SILE.call("par")
  le = string.sub(le, 5, #le)
  SILE.typesetter:typeset(le .. " ‚Üí ")
  if decle(le) == teststr then check() end
  SILE.typesetter:typeset(decle(le))
  end

  test("Schr√∂dinger")
  test("Áå´†Ç§~")

  SILE.call("par")
  local fchecks = checks
  if fchecks == 8 then SILE.typesetter:typeset("OK") else SILE.typesetter:typeset("NG") end
  for i = 1, fchecks do
    check()
  end
  if SILE.backend ~= "debug" then
    require("packages/pdf")
  else
    SILE.registerCommand("pdf:metadata", function (_, _) end)
  end
}
\pdf:metadata[key=Author, value=Erwin Schr√∂dinger]
\pdf:metadata[key=Title, value=Erwin Schr√∂dinger‚Äôs Áå´üêà]
\pdf:metadata[key=CreationDate, value=D:19990209153925 - 08 ' 00 ']
\pdf:metadata[key=ModDate, value=should fail]
\pdf:metadata[key=Trapped, value=should be skipped]
\end{document}
