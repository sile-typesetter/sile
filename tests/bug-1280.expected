Set paper size 	425.196855	170.078742
Begin page
Mx 	21.2598
My 	16.5739
Set font 	;10;400;;normal;;LTR;.fonts/NotoSerifCJK-Regular.ttc
Mx 	5.5700
Mx 	5.1900
Mx 	6.5500
Mx 	4.4000
Mx 	5.9400
Mx 	6.1700
Mx 	3.2500
Mx 	6.5800
Mx 	5.5200
Mx 	5.4900
Mx 	4.5400
T	52 a=5.5700 68 a=5.2900 73 a=6.5500 83 a=4.5400 180 a=5.9400 69 a=6.1700 74 a=3.2500 79 a=6.5800 72 a=5.6000 70 a=5.4900 83 a=4.5400	(Schrödinger)
Mx 	83.5959
T	805 w=10.0000	(⇒)
Mx 	21.2598
My 	28.8939
Mx 	3.6800
Mx 	5.4200
Mx 	7.1400
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	3.8500
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.4900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
T	71 a=3.8500 70 a=5.4900 59871 a=7.1400 62294 a=5.3900 62294 a=5.3900 62299 a=5.3900 62297 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62297 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62302 a=5.3900 62294 a=5.3900 62294 a=5.3900 62301 a=5.3900 62296 a=5.3900 62294 a=5.3900 62294 a=5.3900 71 a=3.8500 62300 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62298 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62303 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 70 a=5.4900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62301 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62299 a=5.3900 62294 a=5.3900 62294 a=5.3900 62301 a=5.3900 62296 a=5.3900	(feff005300630068007200f600640069006e006700650072)
Mx 	276.3221
T	789 w=10.0000	(→)
Mx 	289.4244
T	1322 w=6.4600	(✓)
Mx 	295.8844
Mx 	5.5700
Mx 	5.1900
Mx 	6.5500
Mx 	4.4000
Mx 	5.9400
Mx 	6.1700
Mx 	3.2500
Mx 	6.5800
Mx 	5.5200
Mx 	5.4900
Mx 	4.5400
T	52 a=5.5700 68 a=5.2900 73 a=6.5500 83 a=4.5400 180 a=5.9400 69 a=6.1700 74 a=3.2500 79 a=6.5800 72 a=5.6000 70 a=5.4900 83 a=4.5400	(Schrödinger)
Mx 	21.2598
My 	41.2139
T	62294 62294 62299 62297 62294 62294 62300 62297 62294 62294 62300 62302 62294 62294 62301 62296 62294 62294 71 62300 62294 62294 62300 62298 62294 62294 62300 62303 62294 62294 62300 70 62294 62294 62300 62301 62294 62294 62300 62299 62294 62294 62301 62296 w=235.7200	(005300630068007200f600640069006e006700650072)
Mx 	260.0842
T	789 w=10.0000	(→)
Mx 	273.1886
T	1322 w=6.4600	(✓)
Mx 	279.6486
Mx 	5.5700
Mx 	5.1900
Mx 	6.5500
Mx 	4.4000
Mx 	5.9400
Mx 	6.1700
Mx 	3.2500
Mx 	6.5800
Mx 	5.5200
Mx 	5.4900
Mx 	4.5400
T	52 a=5.5700 68 a=5.2900 73 a=6.5500 83 a=4.5400 180 a=5.9400 69 a=6.1700 74 a=3.2500 79 a=6.5800 72 a=5.6000 70 a=5.4900 83 a=4.5400	(Schrödinger)
Mx 	21.2598
My 	53.5339
Mx 	7.1400
Mx 	3.6800
Mx 	5.4900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	3.8500
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.4900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
T	59871 a=7.1400 71 a=3.8500 70 a=5.4900 62299 a=5.3900 62297 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62297 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62302 a=5.3900 62294 a=5.3900 62294 a=5.3900 62301 a=5.3900 62296 a=5.3900 62294 a=5.3900 62294 a=5.3900 71 a=3.8500 62300 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62298 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62303 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 70 a=5.4900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62301 a=5.3900 62294 a=5.3900 62294 a=5.3900 62300 a=5.3900 62299 a=5.3900 62294 a=5.3900 62294 a=5.3900 62301 a=5.3900 62296 a=5.3900 62294 a=5.3900 62294 a=5.3900	(fffe5300630068007200f600640069006e00670065007200)
Mx 	276.3921
T	789 w=10.0000	(→)
Mx 	289.4944
T	1322 w=6.4600	(✓)
Mx 	295.9544
Mx 	5.5700
Mx 	5.1900
Mx 	6.5500
Mx 	4.4000
Mx 	5.9400
Mx 	6.1700
Mx 	3.2500
Mx 	6.5800
Mx 	5.5200
Mx 	5.4900
Mx 	4.5400
T	52 a=5.5700 68 a=5.2900 73 a=6.5500 83 a=4.5400 180 a=5.9400 69 a=6.1700 74 a=3.2500 79 a=6.5800 72 a=5.6000 70 a=5.4900 83 a=4.5400	(Schrödinger)
Mx 	21.2598
My 	65.8539
T	62299 62297 62294 62294 62300 62297 62294 62294 62300 62302 62294 62294 62301 62296 62294 62294 71 62300 62294 62294 62300 62298 62294 62294 62300 62303 62294 62294 62300 70 62294 62294 62300 62301 62294 62294 62300 62299 62294 62294 62301 62296 62294 62294 w=235.7200	(5300630068007200f600640069006e00670065007200)
Mx 	260.0842
T	789 w=10.0000	(→)
Mx 	273.1886
T	1322 w=6.4600	(✓)
Mx 	279.6486
Mx 	5.5700
Mx 	5.1900
Mx 	6.5500
Mx 	4.4000
Mx 	5.9400
Mx 	6.1700
Mx 	3.2500
Mx 	6.5800
Mx 	5.5200
Mx 	5.4900
Mx 	4.5400
T	52 a=5.5700 68 a=5.2900 73 a=6.5500 83 a=4.5400 180 a=5.9400 69 a=6.1700 74 a=3.2500 79 a=6.5800 72 a=5.6000 70 a=5.4900 83 a=4.5400	(Schrödinger)
Mx 	21.2598
My 	77.8539
T	26728 w=10.0000	(猫)
Mx 	31.2598
T	60429 w=10.0000	(𠂤)
Mx 	41.2598
T	95 w=5.7600	(~)
Mx 	50.1602
T	805 w=10.0000	(⇒)
Mx 	21.2598
My 	89.8539
Mx 	3.6800
Mx 	5.4200
Mx 	7.1400
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	6.2700
Mx 	6.1700
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	6.1700
Mx 	5.2900
Mx 	5.5200
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.4900
T	71 a=3.8500 70 a=5.4900 59871 a=7.1400 62301 a=5.3900 62297 a=5.3900 62296 a=5.3900 67 a=6.2700 69 a=6.1700 62302 a=5.3900 62298 a=5.3900 62294 a=5.3900 69 a=6.1700 68 a=5.2900 66 a=5.5200 62298 a=5.3900 62294 a=5.3900 62294 a=5.3900 62301 a=5.3900 70 a=5.4900	(feff732bd840dca4007e)
Mx 	129.4354
T	789 w=10.0000	(→)
Mx 	142.5610
T	1322 w=6.4600	(✓)
Mx 	149.0210
T	26728 w=10.0000	(猫)
Mx 	159.0210
T	60429 w=10.0000	(𠂤)
Mx 	169.0210
T	95 w=5.7600	(~)
Mx 	21.2598
My 	101.8539
T	62301 62297 62296 67 69 62302 62298 62294 69 68 66 62298 62294 62294 62301 70 w=88.8100	(732bd840dca4007e)
Mx 	113.1975
T	789 w=10.0000	(→)
Mx 	126.3252
T	1322 w=6.4600	(✓)
Mx 	132.7852
T	26728 w=10.0000	(猫)
Mx 	142.7852
T	60429 w=10.0000	(𠂤)
Mx 	152.7852
T	95 w=5.7600	(~)
Mx 	21.2598
My 	113.8539
Mx 	7.1400
Mx 	3.6800
Mx 	5.4900
Mx 	5.3900
Mx 	6.2700
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	5.3900
Mx 	6.1700
Mx 	5.3900
Mx 	5.5200
Mx 	5.3900
Mx 	6.1700
Mx 	5.2900
Mx 	5.3900
Mx 	5.4900
Mx 	5.3900
Mx 	5.3900
T	59871 a=7.1400 71 a=3.8500 70 a=5.4900 62296 a=5.3900 67 a=6.2700 62301 a=5.3900 62297 a=5.3900 62298 a=5.3900 62294 a=5.3900 69 a=6.1700 62302 a=5.3900 66 a=5.5200 62298 a=5.3900 69 a=6.1700 68 a=5.2900 62301 a=5.3900 70 a=5.4900 62294 a=5.3900 62294 a=5.3900	(fffe2b7340d8a4dc7e00)
Mx 	129.5054
T	789 w=10.0000	(→)
Mx 	142.6309
T	1322 w=6.4600	(✓)
Mx 	149.0909
T	26728 w=10.0000	(猫)
Mx 	159.0909
T	60429 w=10.0000	(𠂤)
Mx 	169.0909
T	95 w=5.7600	(~)
Mx 	21.2598
My 	125.8539
T	62296 67 62301 62297 62298 62294 69 62302 66 62298 69 68 62301 70 62294 62294 w=88.8100	(2b7340d8a4dc7e00)
Mx 	113.1975
T	789 w=10.0000	(→)
Mx 	126.3252
T	1322 w=6.4600	(✓)
Mx 	132.7852
T	26728 w=10.0000	(猫)
Mx 	142.7852
T	60429 w=10.0000	(𠂤)
Mx 	152.7852
T	95 w=5.7600	(~)
Mx 	21.2598
My 	137.8539
Mx 	7.4600
Mx 	7.2300
T	48 a=7.6900 44 a=7.2300	(OK)
T	1322 w=6.4600	(✓)
Mx 	42.4098
T	1322 w=6.4600	(✓)
Mx 	48.8698
T	1322 w=6.4600	(✓)
Mx 	55.3298
T	1322 w=6.4600	(✓)
Mx 	61.7898
T	1322 w=6.4600	(✓)
Mx 	68.2498
T	1322 w=6.4600	(✓)
Mx 	74.7098
T	1322 w=6.4600	(✓)
Mx 	81.1698
T	1322 w=6.4600	(✓)
End page
Finish
